### PR TITLE
Generate: Deprecate returning legacy cache by default; Handle `use_cache=False`

### LINKED ISSUE #32843: Cache updating when use_cache = False
### System Info

- `transformers` version: 4.44.0
- Platform: macOS-14.5-arm64-arm-64bit
- Python version: 3.10.0
- Huggingface_hub version: 0.24.5
- Safetensors version: 0.4.4
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.4.0 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: no

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

I'm experimenting with shuffling layers in a pre-trained model. The `layer_idx` inside the Attention object makes this difficult as described in [this issue](https://github.com/huggingface/transformers/issues/28462). To work around this, I'm setting `use_cache = False`, however, even with `use_cache = False`, an error is occurring as `past_key_value.update` is being called in the Attention forward pass. A simple solution would be to use `use_cache` in the forward pass by adding the following `and` logic:

```
if past_key_value is not None and use_cache:
    cache_kwargs = {"sin": sin, "cos": cos}  # Specific to RoPE models
    key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
```

Here is my code to reproduce. The first run through will run because the layers have not switched, but the second run will fail as the cache attempts to update.

```
from transformers import AutoTokenizer, AutoModelForCausalLM
from utils import *
import random
import torch

def run_with_custom_order(model, tokenizer, device, prompts, order):
    original_layers = model.model.layers
    layer_dict = {i: layer for i, layer in enumerate(original_layers)}
    shuffled_layers = torch.nn.ModuleList([layer_dict[i] for i in order])
    model.model.layers = shuffled_layers

    for prompt in prompts:
        inputs = tokenizer(prompt, return_tensors="pt").to(device)
        input_length = inputs["input_ids"].shape[1]
        outputs = model.generate(**inputs, max_new_tokens=20, pad_token_id=tokenizer.eos_token_id, use_cache=False)
        generated_text = tokenizer.decode(outputs[0, input_length:], skip_special_tokens=True)
        print(f"{generated_text}")

    model.model.layers = original_layers
    pass

def main():
    device = (
        "cuda" if torch.cuda.is_available() else 
        "mps" if torch.backends.mps.is_available() else 
        "cpu"
    )    
    llm_name = "microsoft/Phi-3-mini-4k-instruct"
    tokenizer = AutoTokenizer.from_pretrained(llm_name)
    model = AutoModelForCausalLM.from_pretrained(llm_name, torch_dtype=torch.bfloat16, trust_remote_code=True)
    model.to(device)
    model.config.pad_token_id = tokenizer.eos_token_id
    num_layers = len(model.model.layers)

    # Load questions and answers
    dataset_type = "mmlu"  # Change this to "default" for capitals dataset
    num_questions = 1
    questions, _ = load_qa(dataset_type, num_questions)
    prompts = [f"<|user|>\n{question}\nChoose A, B, C, or D:<|end|>\n<|assistant|>" for question in questions]
    
    order = list(range(num_layers))
    run_with_custom_order(model, tokenizer, device, prompts, order)

    random.shuffle(order)
    run_with_custom_order(model, tokenizer, device, prompts, order)

    pass

if __name__ == "__main__":
    main()
```

### Expected behavior

When `use_cache = False`, the cache should not be updating, right?

Happy to help with PRs if you feel its necessary!

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Another step towards using `Cache` everywhere 💪 

This PR makes the following [`Cache`+`generate`]-related changes:
1. Don't initialize a cache when `use_cache=False` (fixes #32843 )
2. `generate` tests now explicitly pass `use_cache`, instead of setting it in `model.config` 🤢 We were relying on a LOT of side effects, and missing the incorrect case mentioned in #32843 
3. Add sanity-checks on cache-related parameters, in `generation_config`
4. Add a deprecation cycle on the default cache return type, so we start returning a `Cache` instance by default on `generate`
5. Isolate all cache initialization logic in `generate` into a single function, and reorganize the logic by blocks

### CODE REVIEW

File: src/transformers/generation/configuration_utils.py [hunk 1]
```diff
@@ -130,9 +130,29 @@ class GenerationConfig(PushToHubMixin):
             [this paper](https://arxiv.org/pdf/1610.02424.pdf) for more details.
         penalty_alpha (`float`, *optional*):
             The values balance the model confidence and the degeneration penalty in contrastive search decoding.
+        dola_layers (`str` or `List[int]`, *optional*):
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:133]: moved up to this documentation section (Parameters that control the generation strategy used), which makes more sense

File: src/transformers/generation/configuration_utils.py [hunk 2]
```diff
@@ -130,9 +130,29 @@ class GenerationConfig(PushToHubMixin):
...
+            The layers to use for DoLa decoding. If `None`, DoLa decoding is not used. If a string, it must
+            be one of "low" or "high", which means using the lower part or higher part of the model layers, respectively.
+            "low" means the first half of the layers up to the first 20 layers, and "high" means the last half of the
+            layers up to the last 20 layers.
+            If a list of integers, it must contain the indices of the layers to use for candidate premature layers in DoLa.
+            The 0-th layer is the word embedding layer of the model. Set to `'low'` to improve long-answer reasoning tasks,
+            `'high'` to improve short-answer tasks. Check the [documentation](https://github.com/huggingface/transformers/blob/main/docs/source/en/generation_strategies.md)
+            or [the paper](https://arxiv.org/abs/2309.03883) for more details.
+
+        > Parameters that control the cache
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:143]: new cache-related docs section in `GenerationConfig`, moved all cache-related flags here

File: src/transformers/generation/configuration_utils.py [hunk 3]
```diff
@@ -544,8 +539,9 @@ def validate(self, is_init=False):
             raise ValueError(f"`max_new_tokens` must be greater than 0, but is {self.max_new_tokens}.")
         if self.pad_token_id is not None and self.pad_token_id < 0:
             warnings.warn(
-                f"`pad_token_id` should be positive but got {self.pad_token_id}. This will cause errors when batch generating, if there is padding. "
-                "Please set `pad_token_id` explicitly by `model.generation_config.pad_token_id=PAD_TOKEN_ID` to avoid errors in generation, and ensure your `input_ids` input does not have negative values."
+                f"`pad_token_id` should be positive but got {self.pad_token_id}. This will cause errors when batch "
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:542]: (>120 chars/line)

File: src/transformers/generation/utils.py [hunk 4]
```diff
@@ -136,27 +136,23 @@ class GenerateDecoderOnlyOutput(ModelOutput):
         sequences (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
             The generated sequences. The second dimension (sequence_length) is either equal to `max_length` or shorter
             if all batches finished early due to the `eos_token_id`.
-        scores (`tuple(torch.FloatTensor)` *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
+        scores (`tuple(torch.FloatTensor)` *optional*, returned when `output_scores=True`):
```
Comment (gante) [inline src/transformers/generation/utils.py:139]: In our docs we often mention that there are two ways to parameterize `generate` (`generation_config` or pass arg to `generate`). I don't think we need to be verbose here.

Also, setting through `config` is deprecated 😉

File: src/transformers/generation/utils.py [hunk 5]
```diff
@@ -136,27 +136,23 @@ class GenerateDecoderOnlyOutput(ModelOutput):
...
-        past_key_values (`tuple(tuple(torch.FloatTensor)))`, *optional*, returned when `use_cache=True` is passed or when `config.use_cache=True`):
-            NOTE: some models have a different `past_key_values` format, confirm with the model's documentation.
-            Usually a Tuple (one element for each layer of the decoder) of tuples (two elements, key tensor and value
-            tensor). The first Tuple is of length `config.n_layers`, with each tuple having 2 tensors of shape
-            `(batch_size, num_heads, sequence_length, embed_size_per_head)`) and optionally if
-            `config.is_encoder_decoder=True` 2 additional tensors of shape `(batch_size, num_heads,
-            encoder_sequence_length, embed_size_per_head)`.
+        past_key_values (`tuple(tuple(torch.FloatTensor)))`, *optional*, returned when `use_cache=True`):
+            Returns the model cache, used to speed up decoding. Different models have a different cache format, check
+            the model's documentation. Usually, a [`~cache_utils.Cache`] instance.
```
Comment (gante) [inline src/transformers/generation/utils.py:155]: rewrote this one. 

The old description was outdated (legacy cache), and we now know that different models have different caches, so we shouldn't be precise here. The model class docs will be more precise, let's redirect users there.

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -328,6 +312,7 @@ class GenerateBeamEncoderDecoderOutput(ModelOutput):
     past_key_values: Optional[Tuple[Tuple[Tuple[torch.FloatTensor]]]] = None
 
 
+# TODO (joao): remove the equivalent classes and typing shortcuts below in v5
```
Comment (gante) [inline src/transformers/generation/utils.py:315]: (these aliases made sense in the past, not anymore)

File: src/transformers/generation/utils.py [hunk 7]
```diff
@@ -1497,6 +1482,127 @@ def _supports_default_dynamic_cache(self) -> bool:
         """
         return self._supports_cache_class and "jamba" not in self.__class__.__name__.lower()
 
+    def _prepare_cache_for_generation(
```
Comment (gante) [inline src/transformers/generation/utils.py:1485]: New function, moving the cache logic from `generate`.

It is doing essentially the same, except for the `Quick escape route 2`, which is new. Added the warning in `Quick escape route 3`.

File: src/transformers/generation/utils.py [hunk 8]
```diff
@@ -2134,11 +2152,33 @@ def typeerror():
                 **model_kwargs,
             )
 
-        # Convert to legacy cache if needed
-        if use_dynamic_cache_by_default and generation_config.return_legacy_cache:
-            if isinstance(result, ModelOutput) and hasattr(result, "past_key_values"):
-                if isinstance(result.past_key_values, (DynamicCache, EncoderDecoderCache)):
-                    result.past_key_values = result.past_key_values.to_legacy_cache()
+        # Convert to legacy cache format if requested
```
Comment (gante) [inline src/transformers/generation/utils.py:2155]: This logic is expanded to handle a deprecation cycle

File: tests/generation/test_utils.py [hunk 9]
```diff
@@ -194,6 +194,7 @@ def _greedy_generate(
         output_attentions=False,
         output_hidden_states=False,
         return_dict_in_generate=False,
+        use_cache=True,
```
Comment (gante) [inline tests/generation/test_utils.py:197]: changes in this file: pass `use_cache` to `generate`, instead of relying on `model.config.use_cache=False` and its side-effects

added a check to confirm that the cache is None when we pass `use_cache=False`

File: src/transformers/generation/configuration_utils.py [hunk 10]
```diff
@@ -702,6 +706,20 @@ def validate(self, is_init=False):
...
+            # In this case, all cache-related arguments should be unset. However, since `use_cache=False` is often used
+            # passed to `generate` directly to hot-fix cache issues, let's raise a warning instead of an error
+            # (otherwise a user might need to overwrite several parameters).
+            no_cache_warning = (
+                "You have set `use_cache` to `False`, but {cache_arg} is set to {cache_arg_value}. {cache_arg} will "
+                "have no effect."
+            )
+            for arg_name in ("cache_implementation", "cache_config", "return_legacy_cache"):
+                if getattr(self, arg_name) is not None:
+                    warnings.warn(
```
Comment (ArthurZucker) [inline src/transformers/generation/configuration_utils.py:719]: would maybe use logger.warn_once

File: src/transformers/generation/utils.py [hunk 11]
```diff
@@ -1497,6 +1482,127 @@ def _supports_default_dynamic_cache(self) -> bool:
...
+        # Quick escape route 1: if the user specifies a cache, we only need to:
+        # a) check for conflicting `generate` arguments
+        # b) convert to the new cache format (if the user passes a legacy cache and model supports it)
+        user_defined_cache = model_kwargs.get(cache_name)
+        if user_defined_cache is not None:
+            if is_torchdynamo_compiling():
+                raise ValueError(
+                    "Passing `past_key_values` is not supported when compiling `model.generate` with torch.compile -- "
+                    "you may get incorrect outputs. Please compile `model.forward` only or use the "
+                    "`cache_implementation` input argument."
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1512]: Unless the cache is static no?

File: src/transformers/generation/utils.py [hunk 12]
```diff
@@ -1497,6 +1482,127 @@ def _supports_default_dynamic_cache(self) -> bool:
...
+                    f"of tuples). `cache_implementation` (set to {generation_config.cache_implementation}) will be "
+                    "ignored.",
+                    UserWarning,
+                )
+            return
+
+        # Otherwise we NEED to prepare a cache, based on `generation_config.cache_implementation`
+
+        # TODO(joao): support static caches in assisted generation. assisted generation needs to roll back caches,
+        # which is only supported in dynamic caches atm
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1546]: let's create an issue and leave it up to the community in the mean time!

**Review by ArthurZucker — APPROVED** [review]
Cool! Let's make sure slow tests all pass as well here!

File: src/transformers/generation/utils.py (hunk 11 above)
Comment (gante) [reply src/transformers/generation/utils.py:1516]: It wasn't working in the PR that added this exception, but I've checked again due to your comment and it seems to be working. Possibly because it now inherits from `nn.module`? 🤷 

Anyways, removing the exception 🤗

File: src/transformers/generation/utils.py (hunk 12 above)
Comment (gante) [reply src/transformers/generation/utils.py:1544]: https://github.com/huggingface/transformers/issues/32946
